text/0021-any-and-all-operators.md

```diff
@@ -0,0 +1,165 @@
+# Add `all` and `any` operators
+
+## Related issues and PRs
+
+- Reference Issues:
+- Implementation PR(s):
+
+## Timeline
+
+- Start Date: 2023-07-14
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+This RFC proposes extending the language with `all` and `any` operators that allow checking if all elements or any element in a set satisfies a given predicate. The `all` operator returns `true` if the provided predicate is true for all elements in the set. The `any` operator returns true if the predicate is true for any element in the set.
+
+## Basic example
+
+Consider writing a policy that allows access when all port numbers in the `context` are between 8000 and 8999.  This can't be expressed in Cedar today.
+
+With the `all` operator, we can write this policy as follows:
+
+```
+permit(principal, action, resource)
+when {
+  context.portNumbers.all(it >= 8000 && it <= 8999)
+};
+```
+
+The same policy can be written using the `any` operator instead:
+
+```
+permit(principal, action, resource)
+unless {
+  context.portNumbers.any(it < 8000 || it > 8999)
+};
+```
+
+The two operators behave analogously to `&&` and `||`. Given negation, each can be written in terms of the other. We include both for ergonomics.
```

Comment:
`any` behaves differently than`||`. `any` does not short-circuit while `||` does. For example. ` 2 > 1  || true > 1` evaluates to `true`, but `[2, true].any(it > 1)` errors.

Reply:
@emina I think @anwarmamat makes an important point here, which is glossed over in the text of the RFC: using `||` and `&&` is order dependent due to short circuiting, whereas use of `any` and `all` cannot short circuit in order to give an order independent result. I'd aim to clarify this point in the RFC text throughout. (There is one place that talks about the use of `contains` as opposed to its encoding using `any`, but this is an indirect statement about the avoidance of short circuiting, not a direct one.)

Reply:
This is explained in detail in the semantics section: https://github.com/cedar-policy/rfcs/blob/any-and-all-operators/text/0021-any-and-all-operators.md#extending-the-semantics

In the case of errors, `.any` and `.all` will error when _any_ expansion to `||` and `&&` errors.  

So, the meaning is "analogous" but not "equivalent".

Reply:
Added a forward pointer to the semantics section in the intro, explicitly stating that the expansion is not equivalent to the use of `||` and `&&`.

Reply:
I had the same question as @anwarmamat here. I think the text could be more clear to callout that the behavior for `any` and `||` is different in the presence of errors.